### PR TITLE
Add mirrord intellij extension

### LINKED ISSUE #24: IntelliJ Extension
We want to have similar functionality and user experience in the IntelliJ extension as we have for VSCode.
We haven't researched how the API looks for IntelliJ but in general it should let the developer run the program with mirrord, and let it select pods in a nice way.

### PR DESCRIPTION
**Pull request by infiniteregrets**
closes #24

### CODE REVIEW

File: intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/ToggleButton.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        return MirrordListener.enabled
+    }
+
+    override fun setSelected(e: AnActionEvent, state: Boolean) {
+        var notificationManager = NotificationGroupManager
+                .getInstance()
+                .getNotificationGroup("Mirrord Notification Handler")
+        if (state) {
+            notificationManager
+                    .createNotification("mirrord", "mirrord configuration is active, current project will run in context of the remote pod when debugged")
```
Comment (eyalb181) [inline intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/ToggleButton.kt:19]: You can just say 'mirrord enabled' and 'mirrord disabled'

File: intellij-ext/src/main/resources/META-INF/plugin.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    <description><![CDATA[
+      mirrord works by letting you select a pod to mirror traffic from.
+      It launches a privileged pod on the same node which enters the namespace of the selected pod and captures traffic from it.
+    ]]></description>
+
+    <depends>com.intellij.modules.lang</depends>
+<!--    <depends>com.intellij.modules.python</depends>-->
+
+    <extensions defaultExtensionNs="com.intellij">
+        <notificationGroup id="Mirrord Notification Handler"
```
Comment (eyalb181) [inline intellij-ext/src/main/resources/META-INF/plugin.xml:16]: Note that mirrord should be all lowercase

File: intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordDialogBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+        dialogPanel.add(checkBoxPanel, BorderLayout.EAST)
+
+        return dialogPanel
+    }
+
+    fun createMirrordNamespaceDialog(namespaces: JBList<String>): JPanel {
+        val dialogPanel = JPanel(BorderLayout())
+        val label = JLabel("Select Namespace to use")
+        val border = EmptyBorder(5, 20, 5, 20)
```
Comment (eyalb181) [inline intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordDialogBuilder.kt:43]: Why not initialize straight into label.border?

File: intellij-ext/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<!-- Keep a Changelog guide -> https://keepachangelog.com -->
```
Comment (eyalb181) [inline intellij-ext/CHANGELOG.md:1]: No need for a separate CHANGELOG for the moment

File: intellij-ext/gradle.properties [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# IntelliJ Platform Artifacts Repositories
+# -> https://plugins.jetbrains.com/docs/intellij/intellij-artifacts.html
+
+pluginGroup = com.github.metalbear.intellijplugin
+pluginName = mirrord-intellij-plugin
+# SemVer format -> https://semver.org
+pluginVersion = 0.0.1
```
Comment (eyalb181) [inline intellij-ext/gradle.properties:7]: Note that version should match the general mirrord version (currently 2.7.0)

**Review by eyalb181 — CHANGES_REQUESTED** [review]
Added a few comments. Also note that a publishing action needs to be added to our CI (see release.yaml)

File: intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordListener.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+class MirrordListener : ExecutionListener {
+    private val mirrordEnv: LinkedHashMap<String, String> = LinkedHashMap()
+    private val log: Logger = Logger.getInstance("MirrordListener")
+
+    init {
+        mirrordEnv["DYLD_INSERT_LIBRARIES"] = "target/debug/libmirrord_layer.dylib"
+        mirrordEnv["LD_PRELOAD"] = "target/debug/libmirrord_layer.so"
+        mirrordEnv["RUST_LOG"] = "DEBUG"
+        mirrordEnv["MIRRORD_AGENT_IMPERSONATED_POD_NAME"] = "nginx-deployment-66b6c48dd5-ggd9n"
```
[the remark this answers is not in the document]
Comment (infiniteregrets) [reply intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordListener.kt]: which part exactly? LD_PRELOAD depends on how we can bundle the binary with the plugin i.e. add it to the plugin zip and i have removed the env that set nginx pod

File: intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordListener.kt [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+                var dialogBuilder = customDialogBuilder.getDialogBuilder(panel)
+
+                // SUCCESS: set the respective environment variables
+                if (dialogBuilder.show() == DialogWrapper.OK_EXIT_CODE) {
+                    mirrordEnv["MIRRORD_AGENT_IMPERSONATED_POD_NAME"] = pods.selectedValue as String
+                    mirrordEnv["MIRRORD_FILE_OPS"] = fileOpsCheckbox.isSelected.toString()
+                    mirrordEnv["MIRRORD_REMOTE_DNS"] = remoteDnsCheckbox.isSelected.toString()
+
+                    var envMap = getPythonEnv(env)
```
[the remark this answers is not in the document]
Comment (infiniteregrets) [reply intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordListener.kt]: yes

File: intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordDialogBuilder.kt (hunk 3 above)
Comment (infiniteregrets) [reply intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordDialogBuilder.kt:59]: iirc setter returns a unit, not this so cant do that here or do you mean something else?

Comment (eyalb181) [reply intellij-ext/src/main/kotlin/com/github/metalbear/intellijplugin/MirrordDialogBuilder.kt:65]: I meant why not just do `label.border = EmptyBorder(5, 20, 5, 20)`?

**Comment by infiniteregrets** [comment]
@meowjesty PTAL

**Review by meowjesty — COMMENTED** [review]
Hmm, the code looks fine. Don't have the kotlin knowledge to tell you where to improve, sorry.

File: intellij-ext/src/main/resources/META-INF/plugin.xml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<!-- Plugin Configuration File. Read more: https://plugins.jetbrains.com/docs/intellij/plugin-configuration-file.html -->
+<idea-plugin>
+    <id>com.github.metalbear.intellijplugin</id>
```
Comment (eyalb181) [inline intellij-ext/src/main/resources/META-INF/plugin.xml:3]: This should probably just be com.github.metalbear.mirrord

**Comment by infiniteregrets** [comment]
Todo(Eyal):
> Anyway bundling both is fine for now but please open an issue to download the correct binary after the plugin was installed

File: intellij-ext/src/main/resources/messages/MyBundle.properties [hunk 9]
```diff
@@ -0,0 +1,3 @@
+name=My Plugin
```
Comment (eyalb181) [inline intellij-ext/src/main/resources/messages/MyBundle.properties:1]: What's this file?

File: intellij-ext/.run/Run IDE for UI Tests.run.xml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+<component name="ProjectRunConfigurationManager">
```
[the remark this answers is not in the document]
Comment (infiniteregrets) [reply intellij-ext/.run/Run IDE for UI Tests.run.xml]: .run is just used by the gradle plugin to call .gradlew buildPlugin or something

File: CHANGELOG.md [hunk 11]
```diff
@@ -9,6 +9,7 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how
 
 ### Added
 - Support for Golang fileops
+- Intellij Extension for mirrord
```
Comment (eyalb181) [inline CHANGELOG.md:12]: Note that it's styled IntelliJ (not too important here, but for future user-facing PRs)

**Review by eyalb181 — APPROVED** [review]
Really great job here @infiniteregrets, well done

**Pull request merged**